### PR TITLE
adds null protection in case of classes have no package.

### PR DESCRIPTION
**Pull request by lordofthejars**
@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java [hunk 1]
```diff
@@ -153,7 +153,19 @@ public boolean permitsMethod(Method method, Object receiver, Object[] args) {
...
+            final Class<?> aClass = receiver.getClass();
+            final Package aPackage = aClass.getPackage();
+
+            if(aPackage == null) {
+                return false;
+            }
+
+            final String name = aPackage.getName();
+
+            if (name == null) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java]: Impossible.

**Comment by jglick** [comment]
:bug: forgot null protection in several other places, such as `permitsConstructor`.

Recommend adding a regression test. Should be easy.

**Comment by lordofthejars** [comment]
I know that there are other places, in fact I try to touch as less lines as possible and I fixes the problem I had now. But now that you mention this I will fix as well.

**Comment by lordofthejars** [comment]
@jglick now I am not sure what you mean by doing a regression test for checking nulls, it comes to my mind a unit test but nothing more

**Comment by jglick** [comment]
> I am not sure what you mean by doing a regression test for checking nulls

Well, you hit an NPE somehow, obviously. So you use `JenkinsRule` to create a `WorkflowJob` that does that same thing, and make sure it builds successfully.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java [hunk 2]
```diff
@@ -163,7 +171,17 @@ public boolean permitsMethod(Method method, Object receiver, Object[] args) {
 
         @Override
         public boolean permitsConstructor(@Nonnull Constructor<?> constructor, @Nonnull Object[] args) {
-            return constructor.getDeclaringClass().getPackage().getName().equals(ORG_APACHE_MAVEN_MODEL);
+            if (constructor == null) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java]: :ant: obviously not possible, given the annotation.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java [hunk 3]
```diff
@@ -173,12 +191,33 @@ public boolean permitsStaticMethod(@Nonnull Method method, @Nonnull Object[] arg
 
         @Override
         public boolean permitsFieldGet(@Nonnull Field field, @Nonnull Object receiver) {
-            return receiver.getClass().getPackage().getName().equals(ORG_APACHE_MAVEN_MODEL);
+
+            if (receiver == null) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java]: :ant: ditto

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java [hunk 4]
```diff
@@ -173,12 +191,33 @@ public boolean permitsStaticMethod(@Nonnull Method method, @Nonnull Object[] arg
...
+            }
+
+            return aPackage.getName().equals(ORG_APACHE_MAVEN_MODEL);
         }
 
         @Override
         public boolean permitsFieldSet(@Nonnull Field field, @Nonnull Object receiver, Object value) {
-            return receiver.getClass().getPackage().getName().equals(ORG_APACHE_MAVEN_MODEL);
+
+            if (receiver == null) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java]: :ant: ditto

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java (hunk 3 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/maven/ReadMavenPomStep.java]: But nonnull is informative, we could receive a null right?

**Comment by rsandell** [comment]
Looks good, just needs a test or two.

**Comment by lordofthejars** [comment]
The problem is that I dont know why the NPE was thrown with a simple read,
I mean it happens but don t know why exactly my simple readManifest fails

**Comment by lordofthejars** [comment]
@rsandell I am trying to reproduce in a simple test the error that a package is null, but it is not possible,I have tried several things like executing code from external groovy script, using `text` or `file` attributes so I don't know exactly when the `permitsMethod` is called with a method without `package`.

**Comment by rsandell** [comment]
ok, thank you for trying. But as Jesse said, you did get the exception somehow in your own pipeline scripts, so it should be reproducible :).

**Pull request merged**